### DISCUSSION TITLE
why-exercise.html

### DISCUSSION THREAD

**Discussion by sunface** [sunface/rust-by-practice#164] [Book Comments]
https://practice.rs/why-exercise.html 

 Learn Rust with Example, Exercise and real Practice, written with ❤️ by https://course.rs team

**Comment by sunface** [comment]
Welcome to practice.rs, please feel free to leave a message if you have any questions  :)

**Reply by lzd358194530** [reply]
star

**Comment by ttoobbyyy** [comment]
good!!!

**Comment by suuzhh** [comment]
it's helpful for me

**Comment by AngleOldPig** [comment]
第1章节的标题关于pracTIce.rs的拼写错了，给朋友截图推荐这里的时候发现的

**Comment by Pichaochang** [comment]
nice, 🐶

**Comment by TomatoDroid** [comment]
必须star

**Comment by MatthiasvB** [comment]
I can run `cargo install mdbook`. It does not create the directory `rust-by-practice`. I create it manually, enter it, and run `mdbook serve`. It does not work, because the directory is empty. Please provide clear instructions on how to get this to work

**Reply by MatthiasvB** [reply]
Okay, installation of `mdbook` is completely separate from actually getting the exercises. You'll find the on https://github.com/sunface/rust-by-practice

**Comment by TQLeung** [comment]
Nice!

**Comment by yaoxingpu** [comment]
sudo git clone git@github.com:sunface/rust-by-practice.git
Cloning into 'rust-by-practice'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

**Reply by bvdbasch** [reply]
Hi, don't know if you've solved this issue already, but the permission error shown to you is because you've cloned the repository with the `sudo` command. This means that the repository was cloned with the super user (`sudo` stands for **s**uper **u**ser **do**), and the permissions of the cloned repository will be set to the super user and not the normal user.

The correct commands would be:
1. `git clone git@github.com:sunface/rust-by-practice.git`
2. `cargo install mdbook`
3. `cd rust-by-practice && mdbook serve en/ `

Hope it helps and have a great day!

**Comment by OnlyAIGC** [comment]
已star，希望老哥可以有个好心情。

**Comment by yjshi2015** [comment]
已star，希望老哥心情更美丽~

**Comment by capybara0-0** [comment]
For me running locally works perfectly as expected.

**Comment by lgln0** [comment]
$ mdbook serve -n 127.0.0.1 zh-CN/
2024-01-10 15:32:32 [INFO] (mdbook::book): Book building has started
2024-01-10 15:32:32 [INFO] (mdbook::book): Running the html backend
2024-01-10 15:32:33 [INFO] (mdbook::cmd::serve): Serving on: http://127.0.0.1:3000
2024-01-10 15:32:33 [INFO] (mdbook::cmd::watch): Listening for changes...
2024-01-10 15:32:33 [INFO] (warp::server): Server::run; addr=127.0.0.1:3000
2024-01-10 15:32:33 [INFO] (warp::server): listening on http://127.0.0.1:3000

之后没反应了，本地运行是怎样的效果？

**Comment by AkshayO3** [comment]
It helps so much,cheers!

**Comment by Mactep1** [comment]
感谢，催更！

**Comment by kl-xiao** [comment]
windows系统下面，mdbook运行不成功

**Reply by kl-xiao** [reply]
PS C:\Users\A23006637\Desktop\py\Rust> cd .\rust-by-practice\
PS C:\Users\A23006637\Desktop\py\Rust\rust-by-practice> mdbook serve .\zh-CN\
2024-04-09 14:57:16 [ERROR] (mdbook::utils): Error: stream did not contain valid UTF-8
PS C:\Users\A23006637\Desktop\py\Rust\rust-by-practice> 
报这个错误

**Comment by sanjeethboddi** [comment]
First encountered rust in 2019. Didn't understand the concepts like borrowing and checker. After 5 years I'm trying to learn rust again. Let's see how it goes.

**Comment by katteXu** [comment]
㖣從我

**Comment by c4nzin** [comment]
thanks!

**Comment by truthixify** [comment]
HELLO, LFG

**Comment by ghost** [comment]
Hi
when I try to run it locally, got following:
Cloning into 'rust-by-practice'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
 
Please make sure you have the correct access rights

**Reply by banesbitt24** [reply]
For anyone else running into this issue, use https://github.com/sunface/rust-by-practice.git

**Comment by ilk-1** [comment]
已 🌟，老哥心情美丽

**Comment by Mayukh-Ghara** [comment]
Let's goo

**Comment by liucanzhu** [comment]
开学

**Comment by grasshausmedia** [comment]
listening and learning about rust for it's diverse utility. thanks for the video.

if anyone wants to collaborate... i made a small rust project to prevent identity theft at pixelmaker.xyz/pear

**Comment by baitangxiaoyu** [comment]
nice

**Comment by Justjustifyjudge** [comment]
感恩大佬
gratefully thanks

**Comment by HugoL0** [comment]
go语言：我尼玛

**Comment by 0x1057** [comment]
I cannot believe I am just finding this...
